src/Orleans/Async/BatchWorker.cs

```diff
@@ -132,5 +132,53 @@ public bool IsIdle()
             else if (waitfortask != null)
                 await waitfortask;
         }
+
+        /// <summary>
+        /// Notify the worker that there is more work, and wait for the current work cycle, and also the next work cycle if there is currently unserviced work.
+        /// </summary>
+        public async Task NotifyAndWaitForWorkToBeServiced()
+        {
+            Task<Task> waitForTaskTask = null;
+            Task waitForTask = null;
+
+            lock (this)
```

Comment:
please avoid lock 'this', 'this' is public so you can't protect it from other locks.  No good way to be confident of avoiding deadlocks.